themes/src/main/resources/theme/keycloak.v2/login/template.ftl

```diff
@@ -221,11 +227,28 @@
         <#nested "socialProviders">
       </footer>
     </main>
+    <div class="pf-v5-l-flex pf-m-column pf-m-gap-lg pf-v5-m-dir-ltr dark-mode-toggle">
+        <label class="pf-v5-c-switch" for="dark-mode-switch" data-ouia-component-type="PF5/Switch" data-ouia-safe="true">
+            <input id="dark-mode-switch" class="pf-v5-c-switch__input" type="checkbox" aria-labelledby="dark-mode-switch-on" aria-label="" x-model="$store.darkMode.on">
+            <span class="pf-v5-c-switch__toggle"></span>
+            <span class="pf-v5-c-switch__label pf-m-on" id="dark-mode-switch-on" aria-hidden="true">🔆</span>
+            <span class="pf-v5-c-switch__label pf-m-off" id="dark-mode-switch-off" aria-hidden="true">🌙</span>
+        </label>
+    </div>
   </div>
 </div>
 <script type="module">
     import Alpine from "alpinejs";
 
+    document.addEventListener('alpine:init', () => {
+        Alpine.store('darkMode', {
+            on: localStorage.getItem('keycloak-darkmode') === null ? window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches : localStorage.getItem('keycloak-darkmode') === 'true',
+
+            set(value) {
+                this.on = value
+            }
+        })
+    })
```

Comment:
**Disclaimer, this is a brain-dump because I took a day to get more familiar with Alpine.js, and as a learning experience I used the code in this PR. So there is a lot of verbatim stuff here that might be interesting for others as well.**

**If you'd like to see it in action, you can check out this [StackBlitz snippet](https://stackblitz.com/edit/vitejs-vite-7j7gml?file=main.js) I created to test this.**

---

If I am [reading the documentation](https://alpinejs.dev/globals/alpine-store#registering-a-store) correctly, we don't have to use this event to bootstrap the store:

> You can either define an Alpine store inside of an `alpine:init` listener (in the case of including Alpine via a `<script>` tag), OR you can define it before manually calling `Alpine.start()` (in the case of importing Alpine into a build):

Although the documentation here refers to a 'build', I believe this also applies to [loading Alpine as a module](https://alpinejs.dev/essentials/installation#as-a-module), as we also have to manually call `Alpine.start()` there. So we can simplify this code to:

```js
Alpine.store('darkMode', {
  on: localStorage.getItem('keycloak-darkmode') === null ? window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches : localStorage.getItem('keycloak-darkmode') === 'true',

  set(value) {
    this.on = value
  },
});
```

I'd also try and extract the logic to determine the dark mode into a separate function for readability, as well as defining some variables for the storage key and media query:

```js
const darkModeKey = 'darkMode';
const darkModeQuery = window.matchMedia('(prefers-color-scheme: dark)');

Alpine.store('darkMode', {
  enabled: isDarkModeEnabled(),
});

function isDarkModeEnabled() {
  if (localStorage.getItem(darkModeKey) === 'true') {
    return true;
  }

  return darkModeQuery.matches;
}
```

There are a couple of changes here to the `isDarkModeEnabled` implementation compared to the original:
- Instead of first checking if the item in local storage is defined by comparing it to `null`, its value is compared to `'true'` immediately, resulting in less conditions.
- The `window.matchMedia` property is no longer feature detected, as it's [widely supported](https://caniuse.com/mdn-api_window_matchmedia) by all browsers on our support target.

Note that I also renamed `on` to `enabled` here, as `on` might be confused with the registration of an event listener (this is just personal preference). 

As I mentioned above, I think it would make sense to move some of the logic that is now in the HTML to this block of JavaScript. Specifically the logic that stores the state in local storage, and the logic that handles changes to the user's preference in dark mode.

We can handle the storage directly when a user takes an explicit action to toggle the state.

```js
Alpine.store('darkMode', {
  enabled: isDarkModeEnabled(),

  toggle() {
    this.enabled = !this.enabled;
    localStorage.setItem(darkModeQuery, this.enabled.toString());
  },
});
```

Ideally the `enabled` property here cannot be written to from the template (e.g. `$store.darkMode.enabled = false`), but is read-only, only setting the state explicitly trough public interfaces such as `toggle()`. To do this, we can use a closure to hold this state outside of the public interface of the store, and use a getter to allow it to be read.

Due to the way Alpine tracks changes to state, we'll have to hold this state in a [reactive object](https://alpinejs.dev/advanced/reactivity#alpine-reactive), so that Alpine can understand when this property is being referenced from the template.

```js
const darkMode = Alpine.reactive({
  value: isDarkModeEnabled(),
});

Alpine.store('darkMode', {
  get enabled() {
    return darkMode.value;
  },

  toggle() {
    darkMode.value = !darkMode.value;
    localStorage.setItem(darkModeKey, darkMode.value.toString());
  },
});
```

We should also be able to [react to changes](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/change_event#example) in the media query as the users chances their preference in real time. So to do that we can start to listen for events when the [store is initialized](https://alpinejs.dev/globals/alpine-store#initializing-stores).

```js
Alpine.store('darkMode', {
  init() {
    darkModeQuery.addEventListener('change', this._determineEnabled.bind(this));
  },

  _determineEnabled() {
    darkMode.value = isDarkModeEnabled();
  },
});
```

Since `darkModeQuery` is a live object, it's `matches` property will always reflect the latest state, so there is no need to re-query using `matchMedia()`, meaning we can call `isDarkModeEnabled()` directly.

Another useful functionality that I'd like to add is the ability to sync the dark mode state between tabs. So if you have multiple tabs open on the same site and switch dark mode on in one of them, it will enable it in all other tabs as well.

We can do this by listening to the [`storage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event) event, which will trigger when a storage object is modified in another tab or window. 

```js
Alpine.store('darkMode', {
  init() {
    window.addEventListener('storage', this._onStorageChange.bind(this));
  },

  _onStorageChange(event) {
    if (event.key !== darkModeKey) {
      return;
    }

    darkMode.value = isDarkModeEnabled();
  },
});
```

If we put all of the aforementioned code together we end up with the following:

```js
const darkModeKey = 'darkMode';
const darkModeQuery = window.matchMedia('(prefers-color-scheme: dark)');
const darkMode = Alpine.reactive({ value: isDarkModeEnabled() });

Alpine.store('darkMode', {
  get enabled() {
    return darkMode.value;
  },

  init() {
    darkModeQuery.addEventListener('change', this._determineEnabled.bind(this));
    window.addEventListener('storage', this._onStorageChange.bind(this));
  },

  toggle() {
    darkMode.value = !darkMode.value;
    localStorage.setItem(darkModeKey, darkMode.value.toString());
  },

  _onStorageChange(event) {
    if (event.key !== darkModeKey) {
      return;
    }

    this._determineEnabled();
  },

  _determineEnabled() {
    darkMode.value = isDarkModeEnabled();
  },
});

function isDarkModeEnabled() {
  if (localStorage.getItem(darkModeKey) === 'true') {
    return true;
  }

  return darkModeQuery.matches;
}
```